src/Utilities/Options/CategorizedAnalyzerConfigOptions.cs

```diff
@@ -0,0 +1,179 @@
+{
+    /// <summary>
+    /// Analyzer configuration options from an .editorconfig file that are parsed into general
+    /// and specific configuration options.
+    /// 
+    /// .editorconfig format:
+    ///  1) General configuration option:
+    ///     (a) "dotnet_code_quality.OptionName = OptionValue"
+    ///  2) Specific configuration option:
+    ///     (a) "dotnet_code_quality.RuleId.OptionName = OptionValue"
+    ///     (b) "dotnet_code_quality.RuleCategory.OptionName = OptionValue"
+    ///    
+    /// .editorconfig examples to configure API surface analyzed by analyzers:
+    ///  1) General configuration option:
+    ///     (a) "dotnet_code_quality.api_surface = all"
+    ///  2) Specific configuration option:
+    ///     (a) "dotnet_code_quality.CA1040.api_surface = public, internal"
+    ///     (b) "dotnet_code_quality.Naming.api_surface = public"
+    ///  See <see cref="SymbolVisibilityGroup"/> for allowed symbol visibility value combinations.
+    /// </summary>
+    internal sealed class CategorizedAnalyzerConfigOptions
+    {
+        private const string KeyPrefix = "dotnet_code_quality.";
+
+        public static readonly CategorizedAnalyzerConfigOptions Empty = new CategorizedAnalyzerConfigOptions(
+            ImmutableDictionary<string, string>.Empty,
+            ImmutableDictionary<string, ImmutableDictionary<string, string>>.Empty);
+
+        private CategorizedAnalyzerConfigOptions(
+            ImmutableDictionary<string, string> generalOptions,
+            ImmutableDictionary<string, ImmutableDictionary<string, string>> specificOptions)
+        {
+            GeneralOptions = generalOptions;
+            SpecificOptions = specificOptions;
+        }
+
+        public ImmutableDictionary<string, string> GeneralOptions { get; }
+        public ImmutableDictionary<string, ImmutableDictionary<string, string>> SpecificOptions { get; }
+
+        public static CategorizedAnalyzerConfigOptions Create(IDictionary<string, string> options)
+        {
+            options = options ?? throw new ArgumentNullException(nameof(options));
+
+            if (options.Count == 0)
+            {
+                return Empty;
+            }
+
+            var generalOptionsBuilder = ImmutableDictionary.CreateBuilder<string, string>(StringComparer.OrdinalIgnoreCase);
+            var specificOptionsBuilder = ImmutableDictionary.CreateBuilder<string, ImmutableDictionary<string, string>.Builder>(StringComparer.OrdinalIgnoreCase);
+            foreach (var kvp in options)
+            {
+                var key = kvp.Key;
+                var value = kvp.Value;
+
+                if (!key.StartsWith(KeyPrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                key = key.Substring(KeyPrefix.Length);
+                var keyParts = key.Split('.').Select(s => s.Trim()).ToImmutableArray();
+                switch (keyParts.Length)
```

Comment:
should it crash on duplicated entry?

Reply:
or duplicated entry is already filtered out?

Reply:
Duplicate keys are overrwritten based on order. At least the current compiler's editorconfig parsing does so, and I have retained that logic in EditorConfigParser.